### DISCUSSION TITLE
Avoid using module globals in doctest

### DISCUSSION THREAD

**Discussion by ydshieh** [pytest-dev/pytest#10277] [Q&A]
When running doctest, a (copy of) the module’s globals will be used, so even without importing a package inside a docstring, the test will pass (if it is imported at the top level of the module). See the example at the end.

However, it implies that when a user copy-paste a docstring to run the example, it could fail due to missing imports, even if the doctest passes in the CI.

In [doctest module](https://docs.python.org/3/library/doctest.html#what-s-the-execution-context), there is a notion called `execution context`, and we can pass `globs=your_dict` to `testmod()` in order to change it.

However, I could not find a way to use such mechanism when launching doctest through `pytest`.

**Question**: Is it possible to use custom globals when running `pytest --doctest-module`?

### Example

```bash
pytest --doctest-modules dummy_test.py -sv --doctest-continue-on-failure
```

**dummy_test.py**
```python
import numpy as np

def foo():
    r"""
    Returns:

    Examples:

    ```python
    >>> c = np.arange(10)

    ```"""
```

**Comment by portyu9** [comment]
Hi @ydshieh 

This behavior comes from Python's `doctest` model rather than pytest specifically.

For doctests found in a Python module, `doctest` uses the module's global namespace. pytest currently follows that behavior, so there is no built-in option equivalent to:

```text
run this doctest with an otherwise empty globals dictionary
```

The `doctest_namespace` fixture goes in the opposite direction: it lets you **add** names to the doctest namespace; it does not remove the module globals.

So if the goal is to guarantee that examples are genuinely copy/pasteable, I would make the imports explicit in the examples:

```python
>>> import package
>>> package.some_function(...)
```

rather than relying on a module-level import that happens to make the doctest pass.

You could build a custom doctest collector/plugin which filters the doctest globals, but that would be custom behavior rather than a supported pytest configuration option, and it would diverge from standard Python `doctest` semantics.

Please mark this answer as accepted if it helped, thank you.
